Handle BoundTypeArguments in the VB operation factory

333fred opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/45382. Fixes https://github.com/dotnet/roslyn/issues/35813.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs for a small review.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -646,6 +648,23 @@ Namespace Microsoft.CodeAnalysis.Operations
+        Private Function CreateBoundTypeArgumentsOperation(boundTypeArguments As BoundTypeArguments) As IInvalidOperation
+            ' This can occur in scenarios involving unresolved extension methods in Strict mode, such as
+            ' element.UnresolvedExtension(Of String)()
+            ' Binding treats the `Of String` as the argument to the method, resulting in one of the children of the BoundBadExpression
```
>Binding treats the `Of String` as the argument to the method, [](start = 14, length = 61)

Is this accurate? This doesn't sound right to me.

AlekseyTs commented on src/Compilers/VisualBasic/Test/IOperation/IOperation/IOperationTests_InvalidExpression.vb:
```diff
@@ -912,5 +912,35 @@ Block[B8] - Exit
+        <Fact, WorkItem(35813, "https://github.com/dotnet/roslyn/issues/35813"), WorkItem(45382, "https://github.com/dotnet/roslyn/issues/45382")>
+        Public Sub InvalidTypeArguments()
+            Dim source = <![CDATA[
+Option Strict On
```
What do we get with option strict off?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -646,6 +648,23 @@ Namespace Microsoft.CodeAnalysis.Operations
         End Function
 
+        Private Function CreateBoundTypeArgumentsOperation(boundTypeArguments As BoundTypeArguments) As IInvalidOperation
+            ' This can occur in scenarios involving unresolved extension methods in Strict mode, such as
```
>This can occur in scenarios involving unresolved extension methods in Strict mode [](start = 14, length = 81)

If the method is no where to be found, why are we assuming it is an extension method?

AlekseyTs commented:
Done with review pass (commit 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [771098483](https://github.com/dotnet/roslyn/pull/50942#issuecomment-771098483) [](ancestors = 771098483)

AlekseyTs commented on src/Compilers/VisualBasic/Test/IOperation/IOperation/IOperationTests_InvalidExpression.vb:
```diff
@@ -912,5 +913,49 @@ Block[B8] - Exit
+      IInvalidOperation (OperationKind.Invalid, Type: ?, IsInvalid) (Syntax: 'node.Extens ... (Of Object)')
+        Children(2):
+            IParameterReferenceOperation: node (OperationKind.ParameterReference, Type: System.Object, IsInvalid) (Syntax: 'node')
+            IInvalidOperation (OperationKind.Invalid, Type: null, IsInvalid) (Syntax: '(Of Object)')
```
>IInvalidOperation (OperationKind.Invalid, Type: null, IsInvalid) (Syntax: '(Of Object)') [](start = 12, length = 88)

Would it make sense to also filter out BoundTypeArguments in CreateBoundBadExpressionOperation?

333fred replied:
I waffled between that approach and creating a node for them, and ended up deciding on the approach I took because it felt like that was more correct: there's a number of odd things that can be the child of a BoundBadExpression, and BoundTypeArguments don't feel special in that regard. That being said, if you feel that we should instead just filter them out I'm not opposed to doing so.

---
In reply to: [568159440](https://github.com/dotnet/roslyn/pull/50942#discussion_r568159440) [](ancestors = 568159440)

AlekseyTs reviewed (approved):
LGTM (commit 2)

333fred commented:
@dotnet/roslyn-compiler for a second review.
